### PR TITLE
Refactor BackendContext to handle nodes request/state + fix restart bug

### PR DESCRIPTION
**Pull request by joeyballentine**
Now we no longer get the backend failed to fetch error during an intentional restart.

### CODE REVIEW

File: src/renderer/contexts/BackendContext.tsx [hunk 1]
```diff
@@ -29,6 +36,8 @@ interface BackendContextState {
     scope: Scope;
     restartingRef: Readonly<React.MutableRefObject<boolean>>;
     restart: () => Promise<void>;
+    nodesQuery: UseQueryResult<BackendNodesResponse, unknown>;
+    nodesInfo: NodesInfo | undefined;
```
Comment (RunDevelopment) [inline src/renderer/contexts/BackendContext.tsx:40]: No. `main.tsx` basically uses these 2 to deduce the current loading state. This is a suuuuper leaky abstraction, or rather no abstraction at all. So let's fix that. Maybe something like `connectionState: "connecting" | "connected" | "failed"` could work.

File: src/renderer/contexts/BackendContext.tsx [hunk 2]
```diff
@@ -37,27 +46,130 @@ export const BackendContext = createContext<Readonly<BackendContextState>>(
...
-    }: React.PropsWithChildren<BackendProviderProps>) => {
+    };
+};
+
+export const BackendProvider = memo(
+    ({ url, pythonInfo, children }: React.PropsWithChildren<BackendProviderProps>) => {
+        const { t } = useTranslation();
         const backend = getBackend(url);
 
+        const { sendAlert, forgetAlert } = useContext(AlertBoxContext);
```
Comment (RunDevelopment) [inline src/renderer/contexts/BackendContext.tsx:77]: I'm not a fan of all this code just existing here. Could we extract all the code you inserted out into a function (hook)? I think it only needs to return `nodesInfo` and the current connection state. Feel free to keep it in this file, I just don't want to make `BackendProvider` this complex.

File: src/renderer/contexts/BackendContext.tsx [hunk 3]
```diff
@@ -120,19 +237,39 @@ export const BackendProvider = memo(
...
+            schemata: nodesInfo?.schemata ?? new SchemaMap([]),
             schemaInputs,
             pythonInfo,
-            categories,
-            categoriesMissingNodes,
-            functionDefinitions,
+            categories: nodesInfo?.categories ?? [],
+            categoriesMissingNodes: nodesInfo?.categoriesMissingNodes ?? [],
+            functionDefinitions:
+                nodesInfo?.functionDefinitions ?? new Map<SchemaId, FunctionDefinition>(),
```
Comment (RunDevelopment) [inline src/renderer/contexts/BackendContext.tsx:267]: You disabled memoizing during loading with those empty arrays and maps. Use `EMPTY_ARRAY` and `EMPTY_MAP` instead. For `SchemaMap` we should probably add something like `SchemaMap.EMPTY`.

File: src/renderer/contexts/BackendContext.tsx [hunk 4]
```diff
@@ -37,25 +46,169 @@ export const BackendContext = createContext<Readonly<BackendContextState>>(
...
+    );
+
+    let connectionState: 'connecting' | 'connected' | 'failed' = 'connecting';
+    if (nodesQuery.status === 'success') {
+        connectionState = 'connected';
+    } else if (nodesQuery.status === 'error') {
+        connectionState = 'failed';
+    }
+
+    return useMemoObject({
```
Comment (RunDevelopment) [inline src/renderer/contexts/BackendContext.tsx:201]: You immediately destructure this object. No need to memo it.

File: src/renderer/main.tsx [hunk 5]
```diff
@@ -207,18 +64,9 @@ export const Main = memo(({ url }: MainProps) => {
...
-            successEffect: setPythonInfo,
-        }),
-        []
-    );
+    if (connectionState === 'failed') return null;
 
-    if (nodesQuery.isError) return null;
-
-    if (!nodesInfo || !pythonInfo || !ready) {
+    if (connectionState === 'connecting' || schemata.schemata.length === 0 || !ready) {
```
Comment (RunDevelopment) [inline src/renderer/main.tsx:69]: Why `schemata.schemata.length === 0`?

Comment (joeyballentine) [reply src/renderer/main.tsx:69]: Is there a better way to check if no nodes are available? If that's not what you mean, I mentioned on discord why I did this.

Comment (RunDevelopment) [reply src/renderer/main.tsx:69]: > it's setting the connectionState to 'connected' before everything is actually defined, so like it connects and goes through everything, but the schemamap is still empty

If you mean this, then I have another question: why did you not fix `connectionState` being `connected` before the data is available? This is a bug in `BackendContext` and shouldn't be fixed here with a work around.

Comment (joeyballentine) [reply src/renderer/main.tsx:69]: ah ok that makes sense

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**